### PR TITLE
Adjust for supporting more implicit ptr-to-int transmutation

### PR DESCRIPTION
**Pull request by RalfJung**
This is the Miri side of https://github.com/rust-lang/rust/pull/101101.
Fixes https://github.com/rust-lang/miri/issues/2456.

### CODE REVIEW

File: src/machine.rs [hunk 1]
```diff
@@ -205,6 +205,23 @@ impl interpret::Provenance for Provenance {
             Provenance::Wildcard => None,
         }
     }
+
+    fn join(left: Option<Self>, right: Option<Self>) -> Option<Self> {
+        match (left, right) {
+            // If one side is `None`, the result is `None`.
+            (None, _) | (_, None) => None,
```
Comment (oli-obk) [inline src/machine.rs:212]: At that point maybe just use

```suggestion
        // If one side is `None`, the result is `None`.
        match (left?, right?) {
```

Comment (RalfJung) [reply src/machine.rs:212]: However, I just realized I can remove this first match arm entirely, and behavior does not even change.

**Comment by RalfJung** [comment]
Hm, some test is taking forever and ui_test doesn't have any kind of timeout...

**Comment by RalfJung** [comment]
It's the `pass/vec.rs` test but I don't know why it times out.

Let's see if a slightly older rustc still works.
@bors r+

**Comment by RalfJung** [comment]
Yeah this times out still...
@bors r- retry

**Comment by RalfJung** [comment]
Blocked on https://github.com/rust-lang/rust/pull/101237.

**Comment by RalfJung** [comment]
Some of the changes here are due to https://github.com/rust-lang/rust/pull/98919, but the conflicts made it too annoying to keep them in a separate commit...

@bors r+

**Pull request merged**